### PR TITLE
feat: use deterministic writeable account order

### PR DESCRIPTION
**Pull request by ChewingGlass**
#### Problem

See https://github.com/solana-labs/solana/issues/21722

#### Summary of Changes

See https://github.com/solana-labs/solana/issues/21722

Fixes #

Fixes ordering of writeable accounts to always be deterministic.

### CODE REVIEW

File: web3.js/src/transaction.ts [hunk 1]
```diff
@@ -261,7 +261,7 @@ export class Transaction {
     accountMetas.sort(function (x, y) {
       const checkSigner = x.isSigner === y.isSigner ? 0 : x.isSigner ? -1 : 1;
       const checkWritable =
-        x.isWritable === y.isWritable ? 0 : x.isWritable ? -1 : 1;
+        x.isWritable === y.isWritable ? x.pubkey.toBuffer().compare(y.pubkey.toBuffer()) : x.isWritable ? -1 : 1;
```
Comment (jstarry) [inline web3.js/src/transaction.ts:264]: We also need this ordering for `checkSigner` right?

Comment (ChewingGlass) [reply web3.js/src/transaction.ts:264]: I thought so, but it causes a whole bunch of tests to fail. I'm not sure if those tests were just hardcoded with an ordering in mind, or there's something more sinister at hand.

Comment (ChewingGlass) [reply web3.js/src/transaction.ts:264]: I don't think so. Just spent a bit trying it out. 

The `checkSigner || checkWritable` means that it falls back to the writable ordering. So writable signers go earlier. If both signers aren't writable, then it uses the fallback buffer sorting.

Comment (jstarry) [reply web3.js/src/transaction.ts:264]: Ah yeah, that's right. Thanks!

**Comment by jstarry** [comment]
Looks good, do you mind fixing the lint issues?

**Comment by ChewingGlass** [comment]
> Looks good, do you mind fixing the lint issues?

Should be good

**Comment by ChewingGlass** [comment]
@jstarry it says first-time contributors need a maintainer to approve running workflows. Lmk if there's anything else I can do on this one

**Comment by jstarry** [comment]
> @jstarry it says first-time contributors need a maintainer to approve running workflows. Lmk if there's anything else I can do on this one

thanks for the ping, just kicked off the job.

**Comment by ChewingGlass** [comment]
@jstarry It looks like some tests are failing. I don't know if that's because the buffers were just manually copy/pasted to compare and the new ones are equivalent, or if there's a bug here. Should I just update the tests with new buffers?

**Comment by ChewingGlass** [comment]
> > @jstarry it says first-time contributors need a maintainer to approve running workflows. Lmk if there's anything else I can do on this one
> 
> thanks for the ping, just kicked off the job.

Needs another workflow run, fixed the issue I think.

**Comment by jstarry** [comment]
thx for your patience!

File: web3.js/src/transaction.ts [hunk 2]
```diff
@@ -259,9 +259,12 @@ export class Transaction {
 
     // Sort. Prioritizing first by signer, then by writable
     accountMetas.sort(function (x, y) {
+      const pubkeySorting = x.pubkey
+        .toBase58()
+        .localeCompare(y.pubkey.toBase58());
```
Comment (mvayngrib) [inline web3.js/src/transaction.ts:264]: can i ask why localeCompare? These are public keys, not sure what internationalization has to do with it? 🤔

Comment (jstarry) [reply web3.js/src/transaction.ts:264]: I didn't notice this in review but agreed that it's overkill here, direct comparison should be fine here. It's fine as is though, imo

Comment (mvayngrib) [reply web3.js/src/transaction.ts:264]: it breaks in react native depending on which JS engine u use, that's how i noticed it. Might be worth refactoring

Comment (jstarry) [reply web3.js/src/transaction.ts:264]: Oh ok, let's refactor then. Do you have the bandwidth to PR the change?

Comment (mvayngrib) [reply web3.js/src/transaction.ts:264]: @jstarry i'll ask someone on my team to upstream the fix 👍

Comment (jstarry) [reply web3.js/src/transaction.ts:264]: Excellent, thanks!

Comment (cassapula) [reply web3.js/src/transaction.ts:264]: @jstarry @mvayngrib apologizes for the delay, but I just created a [PR](https://github.com/solana-labs/solana/pull/28234) to upstream the fix.

**Pull request merged**